### DISCUSSION TITLE
[Portenta H7] UART, SPI, and I2C naming/pins

### DISCUSSION THREAD

**Discussion by darrenweinhold** [micropython/micropython#9603] [Hardware & Peripherals]
I purchased a Portenta H7 and am looking to use it for a project. I ran into a problem with the naming/pinout lineup.

If I look at the pinout page, I see that there are the following UARTs: 1,4, 6, and 8.
https://github.com/micropython/micropython/blob/master/ports/stm32/boards/ARDUINO_PORTENTA_H7/pins.csv

However, if I look at the mpconfigboard.h file, I see the UARTs are: 1, 7, and 8
https://github.com/micropython/micropython/blob/master/ports/stm32/boards/ARDUINO_PORTENTA_H7/mpconfigboard.h

The mpconfigboard.h file corresponds to what I can access in code. This prompts lots of questions. Why is there this difference? How do I know what pins correspond to what UART? Where is the 4th UART (is it for the USB?)?

I have similar questions for I2C and SPI, but I suspect answers surrounding the UART will make things a lot more clear on those as well.
For instance, there should be 3 I2Cs but I only see I2C3. Though I only see I2C1 and I2C3 in the pins file.

@iabdalkader tagging you specifically since I understand you are the author of the board.

**Comment by iabdalkader** [comment]
> If I look at the pinout page, I see that there are the following UARTs: 1,4, 6, and 8.
> https://github.com/micropython/micropython/blob/master/ports/stm32/boards/ARDUINO_PORTENTA_H7/pins.csv

This is not the documentation, the documentation is on the website (https://docs.arduino.cc/hardware/portenta-h7) This file maps named pins to pins. And although it's not wrong (those pins do map to UART AFs) some of them are used for different UARTs and others are on the HDI connector (on the back) which is not normally accessible, or used to connect the vision shield/etc...

> The mpconfigboard.h file corresponds to what I can access in code

Yes that's where you should look and trust for available UARTs, I2Cs etc.. Anything else Not enabled is not enabled for a reason.

**Reply by darrenweinhold** [reply]
Got it. I have a breakout board so am looking to access items that are only available on the high density connectors. I also was not planning on using the vision shield. One of the I2Cs and one of the SPIs are used for that. 

To enable the other UARTs/SPIs/I2Cs, can I just update the mpconfigboard.h file with the appropriate entries and then build the firmware?

**Reply by iabdalkader** [reply]
Yes, but I'd double check the schematics to see if they are used for anything else (like the crypto or the PMIC, I'd avoid those), or if they just connect straight to the HDI.

**Comment by darrenweinhold** [comment]
Update for others: I successfully enabled 7 of the UARTs. The others are not available due to the pin outs of the high density connectors.  

Note: you need to comment out any code relating to the SD Card to use all of the UARTs with my assignments. Check the pins for compatibility! Plenty of things will not function based on these assignments.

Below are my mpconfigboard.h changes to the UART section.

```
// // LPUART config 1
#define MICROPY_HW_LPUART1_TX         (pyb_pin_PB6)
#define MICROPY_HW_LPUART1_RX         (pyb_pin_PB7)

// UART config 1
#define MICROPY_HW_UART1_TX         (pyb_pin_UART1_TX)
#define MICROPY_HW_UART1_RX         (pyb_pin_UART1_RX)
// #define MICROPY_HW_UART_REPL        PYB_UART_1
// #define MICROPY_HW_UART_REPL_BAUD   115200

// UART config 2
#define MICROPY_HW_UART2_TX         (pin_D5)
#define MICROPY_HW_UART2_RX         (pin_D6)

// UART config 4
#define MICROPY_HW_UART4_TX         (pyb_pin_UART4_TX)
#define MICROPY_HW_UART4_RX         (pyb_pin_UART4_RX)

// UART config 6
#define MICROPY_HW_UART6_TX         (pyb_pin_UART6_TX)
#define MICROPY_HW_UART6_RX         (pyb_pin_UART6_RX)

// UART config 7
#define MICROPY_HW_UART7_TX         (pyb_pin_PB3)
#define MICROPY_HW_UART7_RX         (pyb_pin_PB4)

// UART config 8
#define MICROPY_HW_UART8_TX         (pin_J8)
#define MICROPY_HW_UART8_RX         (pin_J9)

```

To enable the LPUART, you need to edit pin_defs_stm32.h. 
Otherwise it won't know what to do with the LPUART. 

`
#define LPUART LPUART
#define GPIO_AF3_LPUART1 GPIO_AF3_LPUART
#define GPIO_AF8_LPUART1 GPIO_AF8_LPUART
`
Then you can use the LP UART in code by calling 
`UART('LP1',9600) `
